scintillometry/channelize/base.py

```diff
@@ -0,0 +1,99 @@
+# Licensed under the GPLv3 - see LICENSE
+
+import operator
+from astropy.utils import lazyproperty
+
+from ..base.base import ModuleBase
+from ..fourier import get_fft_maker
+
+
+__all__ = ['ChannelizeModule']
+
+
+class ChannelizeCore(object):
+
+    # NOTE: maybe force users to pass their own FFT.
+    # NOTE: always assumes time axis is 1.  Probably not a problem.
+    def __init__(self, sample_shape, dtype, nchan, samples_per_block=1,
+                 FFT=None):
+        self.nchan = operator.index(nchan)
+        self.samples_per_block = max(1, operator.index(samples_per_block))
+        FFT = get_fft_maker('numpy') if FFT is None else FFT
+        self._input_shape = (self.samples_per_block, self.nchan) + sample_shape
+        self.fft = FFT(self._input_shape, dtype, axis=1)
+
+    def channelize(self, data):
+        """Split data into frequency channels."""
+        return self.fft(data.reshape(self._input_shape))
+
+
+class ChannelizeModule(ModuleBase, ChannelizeCore):
+
+    def __init__(self, ih, nchan, samples_per_block=1, FFT=None):
+        ModuleBase.__init__(self, ih)
```

Comment:
The original thought was that `ChannelizeCore` could be inherited by both `ChannelizeModule` and `DedisperseModule`, so that the dedispersion didn't have to generate multiple redundant file pointers.  I suppose like with FFTs we could just initialize a channelizer with `ChannelizeModule`, which would help if we ever also implement a polyphase filter instead (though perhaps that would have a larger minimum `samples_per_block`...).

Reply:
Maybe an abstraction to be implemented later? In any case, it still would seem better to use an instance of it rather than inherit from it.

Reply:
Made the change, and I agree it now looks much nicer!